File: Expression.swift
import Foundation

class Expression {
    private var infix: String

    init(_ str: String) { self.infix = str }

    private func priority(_ p: Character, _ q: Character) -> Bool {
        if p == "(" { return false }
        if p == "^" && q == "^" { return false }
        if p == "^" { return true }
        if p == "*" || p == "/" { return true }
        if (p == "+" || p == "-") && (q == "+" || q == "-") { return true }
        return false
    }

    private func infixToPostfix() throws -> [String] {
        var output: [String] = []
        var stack: [Character] = []
        let chars = Array(infix)

        var i = 0
        while i < chars.count {
            let c = chars[i]
            if c.isWhitespace { i += 1; continue }

            if c.isNumber {
                var num = ""
                while i < chars.count, chars[i].isNumber {
                    num.append(chars[i])
                    i += 1
                }
                output.append(num)
                continue
            } else if c == "(" {
                stack.append(c)
            } else if c == ")" {
                while let top = stack.last, top != "(" {
                    output.append(String(stack.removeLast()))
                }
                if stack.isEmpty { throw NSError(domain: "Expression", code: 1, userInfo: [NSLocalizedDescriptionKey: "Unmatched ')'"]) }
                stack.removeLast()
            } else if "+-*/^".contains(c) {
                while let top = stack.last, top != "(", priority(top, c) {
                    output.append(String(stack.removeLast()))
                }
                stack.append(c)
            } else {
                throw NSError(domain: "Expression", code: 2, userInfo: [NSLocalizedDescriptionKey: "Invalid character: \(c)"])
            }
            i += 1
        }

        while !stack.isEmpty {
            let top = stack.removeLast()
            if top != "(" { output.append(String(top)) }
        }

        return output
    }

    func evaluate() throws -> Double {
        let postfix = try infixToPostfix()
        var stack: [Double] = []

        for token in postfix {
            if let num = Double(token) {
                stack.append(num)
            } else if "+-*/^".contains(token) {
                guard stack.count >= 2 else {
                    throw NSError(domain: "Expression", code: 3, userInfo: [NSLocalizedDescriptionKey: "Insufficient operands"])
                }
                let b = stack.removeLast()
                let a = stack.removeLast()
                var res = 0.0
                switch token {
                case "+": res = a + b
                case "-": res = a - b
                case "*": res = a * b
                case "/":
                    guard b != 0 else {
                        throw NSError(domain: "Expression", code: 4, userInfo: [NSLocalizedDescriptionKey: "Division by zero"])
                    }
                    res = a / b
                case "^": res = pow(a, b)
                default: break
                }
                stack.append(res)
            }
        }

        guard stack.count == 1 else {
            throw NSError(domain: "Expression", code: 5, userInfo: [NSLocalizedDescriptionKey: "Invalid expression"])
        }
        return stack[0]
    }
}
extension Expression {
    func prettyFormat() -> String {
        // 将 ^n 替换为上标字符
        let superscriptMap: [Character: Character] = [
            "0": "⁰", "1": "¹", "2": "²", "3": "³", "4": "⁴",
            "5": "⁵", "6": "⁶", "7": "⁷", "8": "⁸", "9": "⁹"
        ]
        
        var output = ""
        var i = infix.startIndex
        while i < infix.endIndex {
            let c = infix[i]
            if c == "^" {
                i = infix.index(after: i)
                while i < infix.endIndex, let mapped = superscriptMap[infix[i]] {
                    output.append(mapped)
                    i = infix.index(after: i)
                }
                continue
            } else {
                output.append(c)
            }
            i = infix.index(after: i)
        }
        return output
    }
}

File: expression.cpp
#include <iostream>
#include <cmath>
#include <string>
#include <stdexcept>
using namespace std;

template<typename T>
struct Stack {
    T data[100];
    int top = -1;

    bool empty() const{return top == -1;}

    void push(T x){
        if(top >= 99) throw runtime_error("Error: Stack overflow");
        data[++top] = x;
    }

    void pop(){
        if(empty()) throw runtime_error("Error: Stack underflow");
        top--;
    }

    T peek()const {
        if(empty()) throw runtime_error("Error: Stack is empty");
        return data[top];
    }

    int size()const {return top + 1;}
};

struct Array {
    string data[200];
    int length = 0;

    void push_back(const string& s) {
        if(length >= 200) throw runtime_error("Error: Array overflow");
        data[length++] = s;
    }
};

bool priority(char p, char q) {
    if(p == '(') return false;
    if(p == '^' && q == '^') return false;
    if(p == '^') return true;
    if(p == '*' || p == '/') return true;
    if((p == '+' || p == '-') && (q == '+' || q == '-')) return true;
    return false;
}

class Expression {
private:
    string infix;

public:
    Expression() = default;
    Expression(string str): infix(str) {}

    void setInfix(string str) { infix = str; }
    string getInfix() const { return infix; }

    Array infixToPostfix() {
        Array outcome;
        Stack<char> S;
        int n = infix.length();

        for(int i = 0; i < n; i++) {
            if(isspace(infix[i])) continue;

            if(isdigit(infix[i])){
                string num;
                while(i < n && isdigit(infix[i])){
                    num += infix[i];
                    i++;
                }
                i--;
                outcome.push_back(num);
            }
            else if(infix[i] == '(') {
                S.push('(');
            }
            else if(infix[i] == ')') {
                while(!S.empty() && S.peek() != '(') {
                    outcome.push_back(string(1, S.peek()));
                    S.pop();
                }
                if(!S.empty()) S.pop();
                else throw runtime_error("Error: Unmatched parenthesis ')'");
            }
            else if(infix[i] == '+' || infix[i] == '-' ||
                     infix[i] == '*' || infix[i] == '/' || infix[i] == '^') {
                while(!S.empty() && S.peek() != '(' && priority(S.peek(), infix[i])){
                    outcome.push_back(string(1, S.peek()));
                    S.pop();
                }
                S.push(infix[i]);
            }
            else{
                throw runtime_error(string("Error: Invalid character '") + infix[i] + "'");
            }
        }

        while(!S.empty()){
            if(S.peek() != '(')
                outcome.push_back(string(1, S.peek()));
            S.pop();
        }
        return outcome;
    }

    double calculatePostfix(const Array& postfix) {
        Stack<double> S;
        for (int i = 0; i < postfix.length; ++i) {
            const string& token = postfix.data[i];
            if(isdigit(token[0])) {
                S.push(stod(token));
            }else{
                if(S.size() < 2)
                    throw runtime_error("Error: Insufficient operands");
                double b = S.peek(); S.pop();
                double a = S.peek(); S.pop();
                double result;
                if(token == "+") result = a + b;
                else if(token == "-") result = a - b;
                else if(token == "*") result = a * b;
                else if(token == "/") {
                    if(b == 0) throw runtime_error("Error: Division by zero");
                    result = a / b;
                }
                else if(token == "^") result = pow(a, b);
                else throw runtime_error("Error: Unknown operator");
                S.push(result);
            }
        }
        if(S.size() != 1) throw runtime_error("Error: Invalid expression");
        return S.peek();
    }

    double evaluate(){
        Array postfix = infixToPostfix();
        return calculatePostfix(postfix);
    }
};